### PR TITLE
Add custom qubit channel 

### PR DESCRIPTION
**Pull request by AroosaIjaz**
Context:

ADR 240: Adding quantum channels to PennyLane

Description of the Change:

This PR adds a custom channel that allows a user to make their own channel and requires a list of Kraus matrices as input. This is an analogous operation to ``qml.QubitUnitary``.

### CODE REVIEW

File: pennylane/ops/channel.py
Comment (antalszava) [inline pennylane/ops/channel.py:39]: My suggestion would be either of the following two:

* Stick with passing lists to QubitChannel and create a new `par_domain` entry if need be: having array nesting `array(array[complex])` becomes complicated and infeasible at times and can have gotchas as the error that arises
* Simply use `return self._kraus_matrices(*self.data)` instead of `return self._kraus_matrices(*self.parameters)` for defining the `kraus_matrices` property for `Channel` --- this disallows training (`self.parameters` can contain `Variable` instances where `self.data` only contains numerical values). Not sure if training based on the input Kraus matrices or any of their parameters would even be a use case

File: pennylane/operation.py
Comment (antalszava) [inline pennylane/operation.py:890]: Will have to stay such that trainable parameters can also be passed
```suggestion
        return self._kraus_matrices(*self.parameters)
```

Comment (AroosaIjaz) [reply pennylane/operation.py:888]: Ah yes! forgot to change this! Thanks!

File: pennylane/ops/channel.py
Comment (antalszava) [inline pennylane/ops/channel.py:40]: Saw that this is identical to the docstring of other ops, would have to be updated 
```suggestion
        wires (Union[Wires, Sequence[int], or int]): the wire(s) the operation acts on
```

File: pennylane/ops/channel.py
Comment (antalszava) [inline pennylane/ops/channel.py:55]: Should there be a check for the dimensions too? Such that the shape is not e.g. `(2,2,2)`.

File: pennylane/operation.py
Comment (antalszava) [inline pennylane/operation.py:469]: A new branch will be needed in the `parameters` property as well for `list` objects. This part failed previously for numpy arrays, now the lists can include numpy arrays which have `Variable` instances in them. This would include something like for the numpy array branch, just iterating through each array:
```python
            if isinstance(p, list):
                # p is assumed to be a list of numpy arrays
                # object arrays may have Variables inside them
                evaled_list = []
                for arr in p:
                    if p.dtype == object:
                        temp.append([x.val if isinstance(x, Variable) else x for x in arr])
                    else:
                        temp.append(p)
                return p
```
Would still be curious to hear @josh146's thoughts if this is going to be fine for trainable parameters.

Comment (AroosaIjaz) [reply pennylane/operation.py:469]: @antalszava should this not go in `evaluate` function ?

File: pennylane/ops/channel.py
Comment (josh146) [inline pennylane/ops/channel.py:69]: This check can be made more efficient by using vectorization rather than list comprehension.

E.g.,
```pycon
>>> k_list = [np.random.random([4, 4]) for _ in range(5)] # 5 2-qubit Kraus matrices
>>> k_arr = np.array(k_list) # convert to array
>>> k_arr.shape
(5, 4, 4)
>>> Kraus_sum = np.einsum("ajk,alk->jl", k_arr, k_arr.conj())
```

**Review by josh146 — COMMENTED** [review]
Looks good to me @AroosaIjaz. Note that I only verified the changes to the operations, and the new Kraus channel class; I didn't review in detail the tests and documentation. Happy for this to be merged once both Antal and Tom approve :)

**Comment by AroosaIjaz** [comment]
@josh146 Can you merge this in too; again waited for codecov for at least a day! Thank you!! :)

**Comment by josh146** [comment]
No worries @AroosaIjaz! However, should we wait for @trbromley's review?

File: pennylane/ops/channel.py
Comment (trbromley) [inline pennylane/ops/channel.py:18]: Why is this line needed?

File: pennylane/ops/channel.py
Comment (trbromley) [inline pennylane/ops/channel.py:256]: Similar, or the same?

File: pennylane/ops/channel.py
Comment (trbromley) [inline pennylane/ops/channel.py:275]: These checks are great, but they seem pretty heavy. From my understanding of how things are working, isn't `_kraus_matrices` called quite regularly? Would it make sense to move these checks to the `__init__` of `QubitChannel`?

File: tests/ops/test_channel_ops.py
Comment (trbromley) [inline tests/ops/test_channel_ops.py:151]:
```suggestion
        # verify equivalent to input matrices
```
?

File: pennylane/operation.py
Comment (AroosaIjaz) [reply pennylane/operation.py:347]: Nice catch! :) I will update it

File: pennylane/ops/channel.py
Comment (AroosaIjaz) [reply pennylane/ops/channel.py:275]: Hm, I don't see how that would be any different.

Comment (trbromley) [reply pennylane/ops/channel.py]: If we place these checks in `QubitChannel.__init__()`, they would just be called once when the channel is instantiated (e.g., a similar thing happens for `PauliRot`)

Instead, they are run in `_kraus_matrices()`, which I believe is called every time the Kraus matrices are requested by the device. This doesn't seem necessary since the kraus matrices are static and only set upon instantiation.

Comment (AroosaIjaz) [reply pennylane/ops/channel.py]: hm nice! I have incorporated this suggestion :)

File: tests/test_operation.py
Comment (trbromley) [inline tests/test_operation.py:137]:
```suggestion
            pars = [[np.eye(2) / np.sqrt(2), np.eye(2) / np.sqrt(2)]] * n
```

**Review by trbromley — APPROVED** [review]
Thanks @AroosaIjaz, looks good! I have left one suggestion to get the tests passing

File: tests/test_operation.py
Comment (AroosaIjaz) [reply tests/test_operation.py]: ah thanks!!

**Comment by AroosaIjaz** [comment]
@josh146 this PR is ready to merge. :)

**Pull request merged**
